pep-0xxx.rst

```diff
@@ -0,0 +1,727 @@
+            bytes, or bytearray value may be supplied directly as the password
+            argument. It will be ignored if the private key is not encrypted
+            and no password is needed.
+            """
+
+        @abstractmethod
+        def set_ciphers(self, ciphers: List[Ciphers]) -> None:
+            """
+            Set the available ciphers for TLS connections created with this
+            context. ``ciphers`` should be a list of ciphers from the
+            ``Cipher`` registry. If none of the ``ciphers`` provided to this
+            object are supported or available, a ``TLSError`` will be raised.
+            """
+
+        @abstractmethod
+        def set_inner_protocols(self, protocols: List[NextProtocol]) -> None:
+            """
+            Specify which protocols the socket should advertise as supported
+            during the TLS handshake. This may be advertised using either or
+            both of ALPN or NPN.
+
+            ``protocols`` should be a list of acceptable protocols in the form
+            of ``NextProtocol`` objects, such as ``[H2, HTTP1]``, ordered by
+            preference. The selection of the protocol will happen during the
+            handshake, and will use whatever protocol negotiation mechanisms
+            are available and supported by both peers.
+
+            If the TLS implementation doesn't support protocol negotiation,
+            this method will raise ``NotImplementedError``.
+            """
+
+        @abstractmethod
+        def set_sni_callback(self, callback: Optional[ServerNameCallback]) -> None:
+            """
+            Register a callback function that will be called after the TLS
+            Client Hello handshake message has been received by the TLS server
+            when the TLS client specifies a server name indication.
+
+            Only one callback can be set per ``Context``. If ``callback`` is
+            ``None`` then the callback is disabled. Calling this function a
+            subsequent time will disable the previously registered callback.
+
+            The ``callback`` function will be called with three arguments: the
+            first will be the ``TLSBufferObject`` for the connection; the
+            second will be a string that represents the server name that the
+            client is intending to communicate (or ``None`` if the TLS Client
+            Hello does not contain a server name); and the third argument will
+            be the original ``Context``. The server name argument will be the
+            IDNA *decoded* server name.
+
+            The ``callback`` must return ``None`` to allow negotiation to
+            continue. Other return values signal errors. Attempting to control
+            what error is signaled by the underlying TLS implementation is not
+            specified in this API, but is up to the concrete implementation to
+            handle.
```

Comment:
I assume the intention here is that the callback does the same `sslbuffer.context` switcheroo as we see in the `ssl` module? If so that should be made explicit (it's not obvious that an implementation needs to support changing `SSLContext` objects midstream!). I do wonder if there's a better option, also (is there any case where the new context differs in any way besides which certificate it presents? could we just like... return the certificate instead?).